packages/blocks-to-image/src/index.ts

```diff
@@ -0,0 +1,146 @@
+import { KnownBlock } from '@slack/types';
+import puppeteer, {
+  BoundingBox,
+  BoxModel,
+  Browser,
+  LaunchOptions,
+  Page,
+  ScreenshotOptions,
+} from 'puppeteer';
+import querystring from 'querystring';
+
+type Screenshot = string | Buffer;
+
+enum BlockSurfaceModeSelectors {
+  'message' = '.p-block_kit_builder_preview__message',
+  'modal' = '.p-block_kit_builder_preview_modal',
+  'appHome' = '.p-block_kit_builder_preview__app_home',
+}
+
+type BlockSurfaceModes = keyof typeof BlockSurfaceModeSelectors;
+
+const minSlackViewportWidth = 1000; // Magic number, width needed for builder to render
+
+async function setViewportFromClip(
+  page: Page,
+  width: number,
+  clip: BoundingBox,
+): Promise<void> {
+  await page.setViewport({
+    width: Math.ceil(width),
+    height: Math.ceil(clip.height + clip.x),
+    deviceScaleFactor: 1,
+  });
+}
+
+async function screenshotDOMElement(
+  page: Page,
+  selector: string,
+  options?: ScreenshotOptions,
+): Promise<Screenshot> {
+  const element = await page.waitForSelector(selector, { visible: true });
+  // Removing the possibility of `null`
+  const elementBox = (await element.boxModel()) as BoxModel;
+
+  const marginTopLeft = elementBox.margin[0];
+  const marginTopRight = elementBox.margin[1];
+  const marginBottomRight = elementBox.margin[3];
+
+  const screenshotOptions = {
+    clip: {
+      x: marginTopLeft.x,
+      y: marginTopLeft.y,
+      width: marginTopRight.x - marginTopLeft.x,
+      height: marginBottomRight.y - marginTopRight.y,
+    },
+  };
+
+  await setViewportFromClip(
+    page,
+    minSlackViewportWidth,
+    screenshotOptions.clip,
+  );
+
+  await page.waitForFunction(
+    `document.querySelector('${selector}') && document.querySelector('${selector}').clientHeight != 0`,
+  );
+
+  const screenshot = await page.screenshot({
+    ...options,
+    ...screenshotOptions,
+  });
+  await page.close();
+
+  return screenshot;
+}
```

Comment:
we could maybe do better function composition here, this is a big function. at the very least, you could break out building the clip from the element:

```ts
async function clipWithMargin(
  element: ElementHandle<Element>,
): Promise<BoundingBox> {
  // Removing the possibility of `null` by waiting for box model of element
  const {
    margin: [
      { x, y }, // marginTopLeft, start x and y
      { x: endX }, // marginTopRight
      ,
      { y: endY }, // marginBottomRight
    ],
  } = (await element.boxModel()) as BoxModel;

  return { x, y, width: endX - x, height: endY - y };
}

async function screenshotDOMElement(
  page: Page,
  selector: string,
  options?: ScreenshotOptions,
): Promise<Screenshot> {
  const element = await page.waitForSelector(selector, { visible: true });
  const clip = await clipWithMargin(element);

  await setViewportFromClip(page, minSlackViewportWidth, clip);

  const screenshot = await element.screenshot({
    ...options,
    clip,
  });
  await page.close();

  return screenshot;
}
```

Reply:
using elementHandle.screenshot made this unnessisary to me